File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, key, val: int):
        self.key = key
        self.val = val
        self.prev = None
        self.head = None 

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}  # values are nodes

        # head is LRU, tail is MRU
        # two dummy nodes at each end
        self.head = Node(-1, 0)
        self.tail = Node(-1, 0)
        self.head.next = self.tail
        self.tail.prev = self.head

    # remove from list but keep cache pointing to it
    def remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

    # add to end of list
    def add(self, node):
        node.prev = self.tail.prev
        node.next = self.tail
        self.tail.prev.next = node
        self.tail.prev = node

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        
        # update order
        self.remove(self.cache[key])
        self.add(self.cache[key])
        return self.cache[key].val

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.remove(self.cache[key])
            self.add(self.cache[key])
            self.cache[key].val = value
            return
        
        if len(self.cache) < self.capacity:
            self.cache[key] = Node(key, value)
            self.add(self.cache[key])
            return

        # evict LRU
        del self.cache[self.head.next.key]
        self.remove(self.head.next)
        self.cache[key] = Node(key, value)
        self.add(self.cache[key])


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.cpp
class LRUCache {
private:
    list<int> order;
    unordered_map<int, pair<int, list<int>::iterator>> cache; 
    int capacity;

public:
    LRUCache(int capacity) {
        this->capacity = capacity;
    }
    
    int get(int key) {
        if (!cache.contains(key))
            return -1;

        order.erase(cache[key].second);
        order.push_back(key);
        cache[key].second = --order.end();
        return cache[key].first;
    }
    
    void put(int key, int value) {
        if (cache.contains(key)) {
            order.erase(cache[key].second);
        } else if (order.size() == capacity) {
            cache.erase(order.front());
            order.pop_front();
        }
        order.push_back(key);
        cache[key] = {value, --order.end()};
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

/*
- doubly linked list to store LRU order
    - node contains value
    - O(1) insertion and removal
- hashmap for store key and index into list
    - O(1) access
*/